File: chroma-league/src/main/java/com/bonepl/chromaleague/hud/parts/resource/GnarFuryBar.java
package com.bonepl.chromaleague.hud.parts.resource;

import com.bonepl.chromaleague.hud.colors.BackgroundBreathingColor;
import com.bonepl.razersdk.color.BreathingColor;
import com.bonepl.razersdk.color.TransitionColor;
import com.bonepl.chromaleague.hud.parts.ProgressBar;
import com.bonepl.chromaleague.state.GameStateHelper;
import com.bonepl.razersdk.animation.AnimatedFrame;
import com.bonepl.razersdk.animation.Frame;
import com.bonepl.razersdk.color.StaticColor;

public class GnarFuryBar extends AnimatedFrame {
    private final TransitionColor fromYellowToRed = new TransitionColor(StaticColor.YELLOW, StaticColor.RED);
    private final BreathingColor aboutToTransform = new BackgroundBreathingColor(StaticColor.YELLOW, 20, false);

    private static final int COLOR_TRANSITION_PERCENT_START = 50;
    private static final int COLOR_TRANSITION_PERCENT_STEP = 100 / COLOR_TRANSITION_PERCENT_START;

    @Override
    public Frame getFrame() {
        final double activePlayerRange = GameStateHelper.getActivePlayerRange();
        final int gnarFuryPercent = GameStateHelper.getResourcePercentage();
        if (activePlayerRange < 300) {
            addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), gnarFuryPercent, StaticColor.RED));
        } else {
            if (gnarFuryPercent < COLOR_TRANSITION_PERCENT_START) {
                addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), gnarFuryPercent, StaticColor.YELLOW));
            } else if (gnarFuryPercent < 85) {
                addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), gnarFuryPercent,
                        fromYellowToRed.getColorAtPercent(COLOR_TRANSITION_PERCENT_STEP * (gnarFuryPercent - COLOR_TRANSITION_PERCENT_START))));
            } else if (gnarFuryPercent < 100) {
                aboutToTransform.setUpColor(fromYellowToRed.getColorAtPercent(COLOR_TRANSITION_PERCENT_STEP * (gnarFuryPercent - COLOR_TRANSITION_PERCENT_START)));
                addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), gnarFuryPercent,
                        aboutToTransform.getColor()));
            } else {
                aboutToTransform.setUpColor(StaticColor.RED);
                aboutToTransform.setSteps(5);
                addAnimationFrame(new ProgressBar(ResourceBars.getResourceBarKeys(), gnarFuryPercent, aboutToTransform.getColor()));
            }
        }
        return super.getFrame();
    }
}

File: src/hud/parts/resource/GnarFuryBar.ts
import {AnimatedFrame} from '../../../razer-sdk/animation/AnimatedFrame.js';
import {Frame} from '../../../razer-sdk/animation/Frame.js';
import {StaticColor} from '../../../razer-sdk/color/StaticColor.js';
import {TransitionColor} from '../../../razer-sdk/color/TransitionColor.js';
import {BackgroundBreathingColor} from '../../colors/BackgroundBreathingColor.js';
import {ProgressBar} from '../ProgressBar.js';
import {getResourceBarKeys} from './ResourceBars.js';
import {getActivePlayerRange, getResourcePercentage} from '../../../state/GameStateHelper.js';

const COLOR_TRANSITION_PERCENT_START = 50;
const COLOR_TRANSITION_PERCENT_STEP = 100 / COLOR_TRANSITION_PERCENT_START;

export class GnarFuryBar extends AnimatedFrame {
    private readonly fromYellowToRed = new TransitionColor(StaticColor.YELLOW, StaticColor.RED);
    private readonly aboutToTransform = new BackgroundBreathingColor(StaticColor.YELLOW, 20, false);

    pollFrame(): Frame {
        const activePlayerRange = getActivePlayerRange();
        const gnarFuryPercent = getResourcePercentage();
        if (activePlayerRange < 300) {
            this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), gnarFuryPercent, StaticColor.RED));
        } else {
            if (gnarFuryPercent < COLOR_TRANSITION_PERCENT_START) {
                this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), gnarFuryPercent, StaticColor.YELLOW));
            } else if (gnarFuryPercent < 85) {
                this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), gnarFuryPercent,
                    this.fromYellowToRed.getColorAtPercent(COLOR_TRANSITION_PERCENT_STEP * (gnarFuryPercent - COLOR_TRANSITION_PERCENT_START))));
            } else if (gnarFuryPercent < 100) {
                this.aboutToTransform.setUpColor(this.fromYellowToRed.getColorAtPercent(COLOR_TRANSITION_PERCENT_STEP * (gnarFuryPercent - COLOR_TRANSITION_PERCENT_START)));
                this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), gnarFuryPercent,
                    this.aboutToTransform));
            } else {
                this.aboutToTransform.setUpColor(StaticColor.RED);
                this.aboutToTransform.setSteps(5);
                this.addAnimationFrame(new ProgressBar(getResourceBarKeys(), gnarFuryPercent, this.aboutToTransform));
            }
        }
        return super.pollFrame();
    }
}
